zookeeper-api/src/test/java/org/apache/helix/zookeeper/impl/client/TestRawZkClient.java

```diff
@@ -285,18 +285,26 @@ public void testZkClientMonitor()
     Assert.assertTrue(beanServer.isRegistered(idealStatename));
 
     Assert.assertEquals((long) beanServer.getAttribute(name, "DataChangeEventCounter"), 0);
-    Assert.assertEquals((long) beanServer.getAttribute(name, "StateChangeEventCounter"), 0);
+    Assert.assertEquals((long) beanServer.getAttribute(name, "StateChangeEventCounter"), 1);
     Assert.assertEquals((long) beanServer.getAttribute(name, "ExpiredSessionCounter"), 0);
     Assert.assertEquals((long) beanServer.getAttribute(name, "OutstandingRequestGauge"), 0);
     // account for doAsyncSync()
     Assert.assertEquals((long) beanServer.getAttribute(name, "TotalCallbackCounter"), 1);
 
-    // Test exists
-    Assert.assertEquals((long) beanServer.getAttribute(rootname, "ReadCounter"), 0);
-    Assert.assertEquals((long) beanServer.getAttribute(rootname, "ReadTotalLatencyCounter"), 0);
-    Assert.assertEquals((long) beanServer.getAttribute(rootname, "ReadLatencyGauge.Max"), 0);
-    zkClient.exists(TEST_ROOT);
+    // Note, we need to wait here for the reason that doAsyncSync() blocks only the zkClient event thread. The main
+    // thread of zkClient would issue exits(TEST_ROOT) without blocking. The return of doAsyncSync() would be asyc
+    // to main thread. doAsyncSync() is a source of 1 read and main thread exists(TEST_ROOT) would be another.
+    TestHelper.verify(()->{
+      return ((org.apache.helix.zookeeper.zkclient.ZkClient)zkClient).getSyncStatus();
+    }, TestHelper.WAIT_DURATION);
+
     Assert.assertEquals((long) beanServer.getAttribute(rootname, "ReadCounter"), 1);
+    //Assert.assertEquals((long) beanServer.getAttribute(rootname, "ReadTotalLatencyCounter"), 1);
```

Comment:
What happen to these 2 checks?

Reply:
Do we really need this check? If we ensure after exists(), the read is 2, then we shall be good, right?
Note that to check this read counter one by one, we need to add one method just for this purpose. I think it is an overkill and it is not necessary.

Reply:
You are right. line 302 and 303 can/should be removed.  The time value is indeterministic. 

line 301 is fine.

Reply:
Seems you have concern about this one by increase it one by one https://github.com/apache/helix/pull/1295/files#r474338328

Let me share my thinking here. 

The doAsyncSync() callback finishing is async to main thread here. It can increase the read counter anytime.  Thus, we need TestHelper.verify to wait for this read from sync(). Here, delete line 301  assert is also ok, just wait for 307 having read counter as 2. Same reason, assert 301 is 1 is fine. 

In sum, either way it is fine. But without TestHelper.verify waiting is not fine.

Reply:
2 ways are different because you will need to add code for supporting the check. If you think both ways of checking are fine, let's do the simpler one.

Reply:
"But without TestHelper.verify waiting is not fine." why is that? You can TestHelper.verify to wait until the count is 2.

Reply:
Unfortunately, you can't or you have race condition in this case, if my understanding is right. Let me illustrate:

doAsycSync(), when the success code returns, it would increase by read counter one from the asyncThread in the ZkClient. 

Note, the main thread running `zkClient.exist(Test_Root )` would increase read counter synchronously from main thread.

The problem is that the doAsynSync() increase to read counter is totally async to the main thread. It is the case that after main thread finish `zkClient.exist(Test_Root )`, the doAsyncSync() would finish for sure. Recent github test actually illustrate this point. Previous we don't realize as it did not happen.

In fact, theoretically, the doAsyncSync() increase can happen a lot later too. 

Let me know if there is other ways you think that we can make the test stable while also no need to use getSyncStatus().

Reply:
Let's sync up tomorrow. I didn't get why waiting for the read count = 2 won't work.

Reply:
changed to waiting for 2